core/src/main/java/org/javarosa/core/util/externalizable/Hasher.java

```diff
@@ -1,20 +1,25 @@
-/**
- *
- */
 package org.javarosa.core.util.externalizable;
 
 /**
- * A class capable of producing hash values. All implementations
- * must provide the same value as the reference implementation
- * (in PrototypeFactory), but can be capable of doing so in
- * an improved manner.
- *
- * This is a workaround for the fact that the serialization engine
- * doesn't provide a PrototypeFactory on serialization, only
- * deserialziation
+ * Abstract hashing class defining the basic outline of performing a hash. Hasher
+ * implementations must override getHash and getHashSize. getClassHashValue() handles
+ * array creation and copying with these methods defined.
  *
  * @author ctsims
+ * @author wspride
  */
-public interface Hasher {
-    public byte[] getClassHashValue(Class type);
+public abstract class Hasher {
+
+    public byte[] getClassHashValue(Class type){
+        byte[] returnHash = new byte[this.getHashSize()];
+        byte[] computedHash = getHash(type); //add support for a salt, in case of collision?
```

Comment:
Ensuring that the hashes will all be of consistent size. I think mutability wouldn't be an issue but a nice safeguard.